crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -0,0 +1,884 @@
+        }
+        Withdraw::Some(withdraw_amount) => {
+            Nominators::<T>::try_mutate(operator_id, nominator_id.clone(), |maybe_nominator| {
+                let nominator = maybe_nominator
+                    .as_mut()
+                    .ok_or(TransitionError::MissingNomination)?;
+
+                // calculate nominator total staked value
+                let nominator_share = Perbill::from_rational(nominator.shares, *total_shares);
+                let nominator_staked_amount = nominator_share.mul_floor(*total_stake);
+
+                // calculate the shares to be deducted from the withdraw amount and adjust
+                let share_per_ssc =
+                    Perbill::from_rational(*total_shares, T::Share::from(*total_stake));
+                let shares_to_withdraw = T::Share::from(share_per_ssc.mul_ceil(withdraw_amount));
+                nominator.shares = nominator
+                    .shares
+                    .checked_sub(&shares_to_withdraw)
+                    .ok_or(TransitionError::ShareUnderflow)?;
+
+                // adjust the locked stake amount with any gains minted
+                let old_locked_amount =
+                    T::Currency::balance_frozen(freeze_identifier, &nominator_id);
+                let amount_to_mint = nominator_staked_amount
+                    .checked_sub(&old_locked_amount)
+                    .unwrap_or(Zero::zero());
+                mint_funds::<T>(&nominator_id, amount_to_mint)?;
+                T::Currency::set_freeze(freeze_identifier, &nominator_id, nominator_staked_amount)
+                    .map_err(|_| TransitionError::UpdateLock)?;
+
+                Ok((withdraw_amount, shares_to_withdraw))
+            })?
+        }
+    };
+
+    PendingNominatorUnlocks::<T>::append(
+        operator_id,
+        unlock_at,
+        PendingNominatorUnlock {
+            nominator_id,
+            balance: withdrew_stake,
+        },
+    );
+
+    let mut operator_ids = PendingUnlocks::<T>::get(unlock_at).unwrap_or_default();
+    operator_ids.insert(operator_id);
+    PendingUnlocks::<T>::insert(unlock_at, operator_ids);
+
+    // update pool's remaining shares and stake
+    *total_shares = total_shares
+        .checked_sub(&withdrew_shares)
+        .ok_or(TransitionError::ShareUnderflow)?;
+    *total_stake = total_stake
+        .checked_sub(&withdrew_stake)
+        .ok_or(TransitionError::BalanceUnderflow)?;
+
+    Ok(())
+}
+
+fn finalize_pending_deposits<T: Config>(
+    operator_id: OperatorId,
+    total_stake: &mut BalanceOf<T>,
+    total_shares: &mut T::Share,
+) -> Result<(), TransitionError> {
+    PendingDeposits::<T>::drain_prefix(operator_id).try_for_each(|(nominator_id, deposit)| {
+        finalize_nominator_deposit::<T>(
+            operator_id,
+            nominator_id,
+            deposit,
+            total_stake,
+            total_shares,
+        )
+    })
+}
+
+fn finalize_nominator_deposit<T: Config>(
```

Comment:
Why not just push this func inside of `finalize_pending_deposits`

Reply:
Easier to read IMO with less and less blocks.